Code/GraphMol/ChemReactions/ReactionRunner.cpp

```diff
@@ -1006,71 +1006,115 @@ void checkAndCorrectChiralityOfMatchingAtomsInProduct(
           !reactant.getBondBetweenAtoms(reactantAtom.getIdx(),
-                                        mapping->prodReactAtomMap[*nbrIdx])) {
+                                        mapping->prodReactAtomMap[nbri])) {
         ++nUnknown;
-        // if there's more than one bond in the product that doesn't correspond
-        // to anything in the reactant, we're also doomed
+        // if there's more than one bond in the product that doesn't
+        // correspond to anything in the reactant, we're also doomed
         if (nUnknown > 1) {
           break;
         }
         // otherwise, add a -1 to the bond order that we'll fill in later
         pOrder.push_back(-1);
       } else {
         const Bond *rBond = reactant.getBondBetweenAtoms(
-            reactantAtom.getIdx(), mapping->prodReactAtomMap[*nbrIdx]);
+            reactantAtom.getIdx(), mapping->prodReactAtomMap[nbri]);
         CHECK_INVARIANT(rBond, "expected reactant bond not found");
         pOrder.push_back(rBond->getIdx());
       }
-      ++nbrIdx;
     }
     if (nUnknown == 1) {
-      // find the reactant bond that hasn't yet been accounted for:
-      int unmatchedBond = -1;
-      boost::tie(nbrIdx, endNbrs) = reactant.getAtomNeighbors(&reactantAtom);
-      while (nbrIdx != endNbrs) {
-        const Bond *rBond =
-            reactant.getBondBetweenAtoms(reactantAtom.getIdx(), *nbrIdx);
-        if (std::find(pOrder.begin(), pOrder.end(), rBond->getIdx()) ==
-            pOrder.end()) {
-          unmatchedBond = rBond->getIdx();
-          break;
+      if (reactantAtom.getDegree() == productAtom->getDegree()) {
+        // there's a reactant bond that hasn't yet been accounted for:
+        int unmatchedBond = -1;
+
+        for (const auto &nbri : boost::make_iterator_range(
+                 reactant.getAtomNeighbors(&reactantAtom))) {
+          const Bond *rBond =
+              reactant.getBondBetweenAtoms(reactantAtom.getIdx(), nbri);
+          if (std::find(pOrder.begin(), pOrder.end(), rBond->getIdx()) ==
+              pOrder.end()) {
+            unmatchedBond = rBond->getIdx();
+            break;
+          }
+        }
+        // what must be true at this point:
+        //  1) there's a -1 in pOrder that we'll substitute for
+        //  2) unmatchedBond contains the index of the substitution
+        auto bPos = std::find(pOrder.begin(), pOrder.end(), -1);
+        if (unmatchedBond >= 0 && bPos != pOrder.end()) {
+          *bPos = unmatchedBond;
+        }
+        if (std::find(pOrder.begin(), pOrder.end(), -1) == pOrder.end()) {
+          nUnknown = 0;
+        }
+      } else if (productAtom->getDegree() > reactantAtom.getDegree()) {
+        // the product has an extra bond. we can just remove the -1 from the
+        // list:
+        auto bPos = std::find(pOrder.begin(), pOrder.end(), -1);
+        pOrder.erase(bPos);
+        if (std::find(pOrder.begin(), pOrder.end(), -1) == pOrder.end()) {
```

Comment:
If I didn't mess up counting the brackets, we got here after checking `nUnknown == 1`, and removing the `-1`. Should we straight out set `nUnknown = 0`, and turn this into an assert ?